Add Total Blocking Time documentation for Lighthouse

kaycebasques opened the pull request:
Fixes #1627 

Changes proposed in this pull request:

- Adds `/src/site/content/en/lighthouse-performance/total-blocking-time/index.md`

Staged page: https://deploy-preview-1636--web-dev-staging.netlify.com/total-blocking-time/

kaycebasques commented:
Hello @patrickhulce and @deepanjanroy,

I wrote documentation for Total Blocking Time. This will be the page that the Lighthouse report UI links to. Please provide a technical review of the content:

- [ ] Is the explanation of what TBT measures accurate?
- [ ] Is the scoring breakdown accurate?
- [ ] Is the guidance relevant?

patrickhulce commented on src/site/content/en/lighthouse-performance/total-blocking-time/index.md:
```diff
@@ -0,0 +1,92 @@
+
+TBT measures the total amount of time that a page is blocked from responding to user input,
+such as mouse clicks, screen taps, or keyboard presses. The sum is calculated by adding
+the duration of all [long tasks] between [First Contentful Paint][fcp] and [Time to Interactive][tti].
```
`[long tasks]` missing a link?

technically, it's summing all of the *blocking portion* of all long tasks between FCP and TTI, i.e. how much longer than 50ms was each long task. not sure how to word that eloquently though

kaycebasques replied:
Actually this is fancy shorthand syntax! Similar to object shorthand `{name}` in JS. Surma used this in a post a few weeks back and I had the exact same response as you.

kaycebasques replied:
Glad I asked for your review. I didn't catch that *blocking portion* stuff. When in doubt, I just add an example to explain complicated concepts. If the updated text looks good to you please resolve this convo.

patrickhulce replied:
woah neat well TIL! and cool this looks great 👍

patrickhulce commented on src/site/content/en/lighthouse-performance/total-blocking-time/index.md:
```diff
@@ -0,0 +1,92 @@
+    </thead>
+    <tbody>
+      <tr>
+        <td>0–200</td>
```
green starts at around ~300, the PODR in the source is roughly a score of 98 :)

kaycebasques replied:
Done

patrickhulce commented on src/site/content/en/lighthouse-performance/total-blocking-time/index.md:
```diff
@@ -0,0 +1,92 @@
+* Inefficient JavaScript statements. For example, after analyzing your code in the Performance panel, suppose
+  you see a call to `document.querySelectorAll('a')` that returns 2000 nodes. Refactoring your code to
+  use a more specific selector that only returns 10 nodes should improve your TBT score.
+* Unnecessary JavaScript loading, parsing, or execution. While analyzing your code in the Performance panel
```
can we move this one first? :) in my experience this is overwhelmingly the more common culprit and yet folks tend to over-obsess about minor js inefficiencies in their code instead

kaycebasques replied:
Definitely. I also added an `aside` to make this point explicit

patrickhulce commented on src/site/content/en/lighthouse-performance/total-blocking-time/index.md:
```diff
@@ -0,0 +1,92 @@
+
+## Resources
+
+- [Source code for **Total Blocking Time** audit](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/metrics/total-blocking-time.js)
```
not sure if we want to fix these to a particular version instead of `master` in case of file moves or whatnot, I don't expect this file to move within the next year or two, but ya never know :)

kaycebasques replied:
I'm working on a content bot that will detect discrepancies between the source and the docs, which should address your point, so I think pointing to master should be fine

patrickhulce commented on src/site/content/en/lighthouse-performance/total-blocking-time/index.md:
```diff
@@ -0,0 +1,92 @@
+In general, the most common causes of long tasks are:
+
+* Inefficient JavaScript statements. For example, after analyzing your code in the Performance panel, suppose
+  you see a call to `document.querySelectorAll('a')` that returns 2000 nodes. Refactoring your code to
```
this specific example of the nodes is something I'm not super confident is common, maybe we could replace with a reference to force reflows or something instead? i.e. "instead of setting style and measuring in a loop over all 2000 nodes, measure all nodes once and then set the styles second" or something

https://developers.google.com/web/fundamentals/performance/rendering/avoid-large-complex-layouts-and-layout-thrashing#avoid_layout_thrashing perhaps?

kaycebasques replied:
I got the querySelectorAll example from Addy's post: https://web.dev/long-tasks-devtools/#what-is-causing-my-long-tasks

I can change this if you feel strongly about it but I think the simple example gets the idea across. Per your other comment this also doesn't seem like a big topic to focus on (but we should probably still mention just in case).

patrickhulce replied:
Ah OK, no worries this works then the simplicity *is* nice! My other comment was the more important piece of this deal :)

kaycebasques commented:
@patrickhulce thank you for the speedy review, please go through your previous comments and resolve them if you think the updated content addresses your feedback

patrickhulce commented:
My resolve comment button seems to be missing on this repo, but everything looks addressed to me!

deepanjanroy commented on src/site/content/en/lighthouse-performance/total-blocking-time/index.md:
```diff
@@ -0,0 +1,99 @@
+the *blocking portion* of all [long tasks] between [First Contentful Paint][fcp] and [Time to Interactive][tti].
+Any task that executes for more than 50&nbsp;ms is a long task. The amount of time after
+50&nbsp;ms is the blocking portion. For example, if Lighthouse detects a 70&nbsp;ms long task,
+the blocking portion would be 20&nbsp;ms.
```
This is very clear! Thank you for this.

kaycebasques commented:
Thank you @patrickhulce and @deepanjanroy!

kaycebasques commented:
@robdodson good point. I renamed it to `lighthouse-total-blocking-time`. This is the pattern that you and I agree to use for Largest Contentful Paint. In the long run I'd prefer to have `lighthouse/total-blocking-time` and use `lighthouse/` for all the audit docs. It'll be tricky to redirect the existing Lighthouse docs, like `/first-contentful-paint`, to Phil's docs now that Lighthouse is linking to web.dev.

kaycebasques commented:
I reviewed this with a fresh pair of eyes. Since Rob also reviewed, as well as Patrick (who's usually pretty thorough even though he's only responsible for a technical review) the content should be fine, so I'm going to merge.

robdodson commented:
> @robdodson good point. I renamed it to lighthouse-total-blocking-time. This is the pattern that you and I agreed to use for Largest Contentful Paint. In the long run I'd prefer to have lighthouse/total-blocking-time and use lighthouse/ for all the audit docs. It'll be tricky to redirect the existing Lighthouse docs, like /first-contentful-paint, to Phil's docs now that Lighthouse is linking to web.dev. We could use some other keyword for Phil's docs, e.g. total-blocking-time-guide.

Another option would be to have Phil's guides use something like `understanding-largest-contentful-paint`.
